src/OrleansRuntime/Scheduler/ShedulerExtensions.cs

```diff
@@ -7,64 +7,23 @@ internal static class SchedulerExtensions
     {
         internal static Task<T> QueueTask<T>(this OrleansTaskScheduler scheduler, Func<Task<T>> taskFunc, ISchedulingContext targetContext)
         {
-            var resolver = new TaskCompletionSource<T>();
-            Func<Task> asyncFunc =
-                async () =>
-                {
-                    try
-                    {
-                        T result = await taskFunc();
-                        resolver.TrySetResult(result);
-                    }
-                    catch (Exception exc)
-                    {
-                        resolver.TrySetException(exc);
-                    }
-                };
-
-            // it appears that it's not important that we fire-and-forget asyncFunc() because we wait on the
-            scheduler.QueueWorkItem(new ClosureWorkItem(() => asyncFunc().Ignore()), targetContext);
-            return resolver.Task;
+            var workItem = new AsyncClosureWorkItem<T>(taskFunc);
+            scheduler.QueueWorkItem(workItem, targetContext);
+            return workItem.Task;
         }
 
         internal static Task QueueTask(this OrleansTaskScheduler scheduler, Func<Task> taskFunc, ISchedulingContext targetContext)
         {
-            var resolver = new TaskCompletionSource<bool>();
-            Func<Task> asyncFunc =
-                async () =>
-                {
-                    try
-                    {
-                        await taskFunc();
-                        resolver.TrySetResult(true);
-                    }
-                    catch (Exception exc)
-                    {
-                        resolver.TrySetException(exc);
-                    }
-                };
-            scheduler.QueueWorkItem(new ClosureWorkItem(() => asyncFunc().Ignore()), targetContext);
-            return resolver.Task;
+            var workItem = new AsyncClosureWorkItem(taskFunc);
+            scheduler.QueueWorkItem(workItem, targetContext);
+            return workItem.Task;
         }
 
         internal static Task QueueNamedTask(this OrleansTaskScheduler scheduler, Func<Task> taskFunc, ISchedulingContext targetContext, string activityName = null)
         {
-            var resolver = new TaskCompletionSource<bool>();
-            Func<Task> asyncFunc =
-                async () =>
-                {
-                    try
-                    {
-                        await taskFunc();
-                        resolver.TrySetResult(true);
-                    }
-                    catch (Exception exc)
-                    {
-                        resolver.TrySetException(exc);
-                    }
-                };
-            scheduler.QueueWorkItem(new ClosureWorkItem(() => asyncFunc().Ignore(), () => activityName), targetContext);
```

Comment:
Sorry, Gabi, I misunderstood. I'm only changing `ClosureWorkItem` and the other ctor calls all pass a const string. `InvokeWorkItem`, etc, can keep doing what they're doing. I didn't see a good reason to change those (although I see a curiosity in the way `OnActivationCompletedRequest` is called in the catch block: it seems it should be scheduled rather than run inline.)